presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java

```diff
@@ -0,0 +1,700 @@
+    private int[] elementPositions;         // positions in elementStream corresponding to positions passed to read(); relative to elementReadOffset
+    private int elementOutputPositionCount;
+    private int[] elementOutputPositions;
+
+    private boolean valuesInUse;
+
+    public ListSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Map<Subfield, TupleDomainFilter> filters,
+            List<Subfield> subfields,
+            ListFilter listFilter,
+            int subfieldLevel,  // 0 - top level
+            Optional<Type> outputType,
+            DateTimeZone hiveStorageTimeZone,
+            AggregatedMemoryContext systemMemoryContext)
+    {
+        requireNonNull(filters, "filters is null");
+        requireNonNull(subfields, "subfields is null");
+        requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+
+        checkArgument(subfields.isEmpty(), "Subfield pruning is not supported yet");
+
+        if (listFilter != null) {
+            checkArgument(subfieldLevel > 0, "SubfieldFilter is not expected at the top level");
+            checkArgument(filters.isEmpty(), "Range filters are not expected at mid level");
+        }
+
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.outputRequired = requireNonNull(outputType, "outputType is null").isPresent();
+        this.outputType = outputType.orElse(null);
+        this.level = subfieldLevel;
+
+        if (listFilter != null) {
+            nullsAllowed = true;
+            nonNullsAllowed = true;
+            this.listFilter = listFilter;
+            nullsFilter = listFilter.getParent().getNullsFilter();
+        }
+        else if (!filters.isEmpty()) {
+            Optional<TupleDomainFilter> topLevelFilter = getTopLevelFilter(filters);
+            if (topLevelFilter.isPresent()) {
+                nullsAllowed = topLevelFilter.get() == IS_NULL;
+                nonNullsAllowed = !nullsAllowed;
+            }
+            else {
+                nullsAllowed = filters.values().stream().allMatch(TupleDomainFilter::testNull);
+                nonNullsAllowed = true;
+            }
+
+            if (filters.keySet().stream().anyMatch(path -> !path.getPath().isEmpty())) {
+                this.listFilter = new ListFilter(streamDescriptor, filters);
+            }
+            else {
+                this.listFilter = null;
+            }
+
+            nullsFilter = null;
+        }
+        else {
+            nullsAllowed = true;
+            nonNullsAllowed = true;
+            this.listFilter = null;
+            nullsFilter = null;
+        }
+
+        StreamDescriptor elementStreamDescriptor = streamDescriptor.getNestedStreams().get(0);
+        Optional<Type> elementOutputType = outputType.map(type -> type.getTypeParameters().get(0));
+
+        switch (elementStreamDescriptor.getStreamType()) {
```

Comment:
Can we reuse `createStreamReader` instead of having a switch? The `switch` is kinda duplicated in both code paths.